src/dfcx_scrapi/tools_ml/utterance_generator_utils.py

```diff
@@ -0,0 +1,369 @@
+
+    def create_new_training_phrases(
+        self, 
+        agent_id: str, 
+        intent_subset: List[str], 
+        new_phrases: int = 100
+    ) -> pd.DataFrame:
+        """ Creates new training phrases for a given list of intents. 
+        
+        The phrases in this set will not be exact string match phrases which
+        exist in the training phrases but will be close semantically. This can
+        be used to run through the core.intents modify_training_phrase_df
+        function to create a new training phrase dataframe. This and a
+        parameters dataframe can be run through tools.dataframe_functions
+        bulk_update_intents_from_dataframe function to make the updates in a
+        dialogflow agent. The new_training output dataframe can be used as the
+        input actions dataframe in the bulk_update_intents_from_dataframe
+        function.
+
+        Args:
+            agent_id: name parameter of the agent to pull intents from - full
+                path to agent.
+            intent_subset: display names of the intents to create a new phrases
+                for, base phrases come from the training in the intent.
+            new_phrases: overall target size of new phrases to create, may be
+                less depending if new independent phrases can be generated from
+                the data. The function tries to get even entries per intent.
+
+        Returns:
+            Dataframe with columns:
+                display_name: intent to add the phrase to.
+                phrase: new phrase. 
+                action: "add".
+        """
+        synthetic_dataset = self.create_synthetic_dataset(
+            agent_id, intent_subset, new_phrases
+        )
+        new_training = (
+            synthetic_dataset.copy()[["intent", "synthetic_phrases"]]
+            .rename(columns={
+                "intent": "display_name", "synthetic_phrases": "phrase"})
+            .reset_index(drop=True)
+        )
+        new_training.insert(len(new_training.columns), "action", "add")
+        return new_training
+
+    @staticmethod
+    def progress_bar(
+        current: int, 
+        total: int, 
+        bar_length: int = 50, 
+        type_: str = "Progress"
+    ):
+        """Display progress bar for processing.
+        
+        Args: 
+            current: number for current iteration.
+            total: number for total iterations.
+            bar_length: number of spaces to make the progress bar,
+                default 50.
+            type_: label for the bar, default 'Progress'.
+        """
+        percent = float(current) * 100 / total
+        arrow = "-" * int(percent / 100 * bar_length - 1) + ">"
+        spaces = " " * (bar_length - len(arrow))
+        print(
+            "{2}({0}/{1})".format(current, total, type_)
+            + "[%s%s] %d %%" % (arrow, spaces, percent),
+            end="\r",
+        )
+
+    @staticmethod
+    def clean_string(string_raw: str) -> str:
```

Comment:
@Hgithubacct Should this be a private function?

Reply:
ack. updated.